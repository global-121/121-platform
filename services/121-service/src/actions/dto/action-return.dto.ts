import {
  ActionArray,
  ActionType,
} from '@121-service/src/actions/action.entity';
import { UserOwnerDto } from '@121-service/src/user/dto/user-owner.dto';
import { ApiProperty } from '@nestjs/swagger';

export class ActionReturnDto {
  @ApiProperty({ example: 1, type: 'number' })
  id: number;
  @ApiProperty({ example: 1, enum: ActionArray })
  actionType: ActionType;
  @ApiProperty({ type: () => UserOwnerDto })
  user: UserOwnerDto;
  @ApiProperty({ example: '2021-09-01T00:00:00.000Z' })
  created: Date;
}
