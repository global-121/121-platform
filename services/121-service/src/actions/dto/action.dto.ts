import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsNumber, IsIn } from 'class-validator';
import { ActionType, ActionArray } from '../action.entity';

export class ActionDto {
  @ApiProperty({
    enum: ActionArray,
    example: ActionArray.join(' | '),
  })
  @IsNotEmpty()
  @IsIn(ActionArray)
  public readonly actionType: ActionType;
}
