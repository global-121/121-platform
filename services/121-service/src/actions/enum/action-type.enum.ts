export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
