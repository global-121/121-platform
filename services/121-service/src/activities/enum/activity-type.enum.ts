export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
