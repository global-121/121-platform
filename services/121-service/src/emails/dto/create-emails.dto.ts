export interface CreateUserEmailPayload {
  email: string;
  displayName: string;
  password?: string;
}

export interface GenericEmailPayload {
  email: string;
  subject: string;
  body: string;
}
