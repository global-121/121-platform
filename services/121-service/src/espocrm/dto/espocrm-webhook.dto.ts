import { ApiProperty } from '@nestjs/swagger';
import { IsEnum, IsString } from 'class-validator';
import { EspoCrmActionTypeEnum } from '../espocrm-action-type.enum';
import { EspoCrmEntityTypeEnum } from './../espocrm-entity-type';

export class EspocrmWebhookDto {
  @ApiProperty({
    description:
      'ID of the EspoCRM webhook entity. This can be copied from the URL when editing the webhook in EspoCRM.',
    example: '63f77488410458465',
  })
  @IsString()
  public referenceId: string;

  @ApiProperty({
    description: 'The type of action that triggers this webhook.',
    enum: EspoCrmActionTypeEnum,
    example: EspoCrmActionTypeEnum.update,
  })
  @IsEnum(EspoCrmActionTypeEnum)
  public actionType: EspoCrmActionTypeEnum;

  @ApiProperty({
    description: 'The entity type that triggers this webhook.',
    enum: EspoCrmEntityTypeEnum,
    example: EspoCrmEntityTypeEnum.registration,
  })
  @IsEnum(EspoCrmEntityTypeEnum)
  public entityType: EspoCrmEntityTypeEnum;

  @ApiProperty({
    description: 'The secret key that is used to verify the webhook.',
    example: 'secret-key',
  })
  @IsString()
  public secretKey: string;
}
