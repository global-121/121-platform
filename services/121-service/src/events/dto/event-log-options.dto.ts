import { EventAttributeKeyEnum } from '@121-service/src/events/enum/event-attribute-key.enum';

/**
 * Data Transfer Object for event log options.
 */
export class EventLogOptionsDto {
  /**
   * An array of registration attributes which will be logged if they have been changed. The default is to check all registration attributes.
   * This is required when using partial RegistrationViewEntities
   */
  registrationAttributes?: string[];

  /**
   * An object of additional log attributes, for example the reason
   * The keys are attribute names and the values are attribute values.
   */
  additionalLogAttributes?: {
    [K in EventAttributeKeyEnum]?: string | undefined;
  };
}
