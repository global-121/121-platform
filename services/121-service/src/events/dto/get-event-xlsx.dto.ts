import { EventEnum } from '@121-service/src/events/enum/event.enum';
import { EventAttributeKeyEnum } from '@121-service/src/events/enum/event-attribute-key.enum';

export class GetEventXlsxDto {
  paId: number;
  referenceId: string;
  type: EventEnum;
  changedBy: string;
  changedAt: Date;

  [EventAttributeKeyEnum.reason]?: string;
  [EventAttributeKeyEnum.oldValue]?: string;
  [EventAttributeKeyEnum.newValue]?: string;
  [EventAttributeKeyEnum.fieldName]?: string;
}
