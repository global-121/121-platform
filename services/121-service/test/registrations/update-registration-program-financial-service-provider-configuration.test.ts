import { HttpStatus } from '@nestjs/common';

import {
  FinancialServiceProviderConfigurationProperties,
  FinancialServiceProviders,
} from '@121-service/src/financial-service-providers/enum/financial-service-provider-name.enum';
import { CreateProgramFinancialServiceProviderConfigurationDto } from '@121-service/src/program-financial-service-provider-configurations/dtos/create-program-financial-service-provider-configuration.dto';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import { registrationVisa } from '@121-service/src/seed-data/mock/visa-card.data';
import { PermissionEnum } from '@121-service/src/user/enum/permission.enum';
import { DefaultUserRole } from '@121-service/src/user/user-role.enum';
import {
  doPayment,
  getTransactions,
  waitForPaymentTransactionsToComplete,
} from '@121-service/test/helpers/program.helper';
import { postProgramFinancialServiceProviderConfiguration } from '@121-service/test/helpers/program-financial-service-provider-configuration.helper';
import {
  importRegistrations,
  seedIncludedRegistrations,
  updateRegistration,
} from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  removePermissionsFromRole,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import { registrationPvScoped } from '@121-service/test/registrations/pagination/pagination-data';

const programIdPv = 2;
const programIdOcw = 3;

async function setupNlrcEnvironment() {
  await resetDB(SeedScript.nlrcMultiple);
  const accessToken = await getAccessToken();

  await importRegistrations(programIdOcw, [registrationVisa], accessToken);
  await importRegistrations(programIdPv, [registrationPvScoped], accessToken);

  return accessToken;
}

describe('Update program financial servce provider configuration of PA', () => {
  let accessToken: string;

  beforeEach(async () => {
    accessToken = await setupNlrcEnvironment();
  });

  it('should succeed when updating program financial servce provider configuration when all required properties of new FSP are already present', async () => {
    // Arrange
    await setupNlrcEnvironment();

    // Intersolve-visa and Intersolve-voucher-whatsapp both have whatsappPhoneNumber as required
    const newProgramFinancialServiceProviderConfigurationName =
      'Intersolve-voucher-whatsapp';
    const dataUpdate = {
      programFinancialServiceProviderConfigurationName:
        newProgramFinancialServiceProviderConfigurationName,
    };
    const reason = 'automated test';

    // Act
    const response = await updateRegistration(
      programIdOcw,
      registrationVisa.referenceId,
      dataUpdate,
      reason,
      accessToken,
    );

    // Assert
    expect(response.statusCode).toBe(HttpStatus.OK);
    expect(response.body.financialServiceProviderName).toBe(
      newProgramFinancialServiceProviderConfigurationName,
    );
  });

  it('should fail when updating program financial servce provider configuration when a required property of new FSP is not yet present', async () => {
    // Arrange
    await setupNlrcEnvironment();

    // Intersolve-voucher-whtatsapp does not have e.g. addressStreet which is required for Intersolve-visa
    const newProgramFinancialServiceProviderConfigurationName =
      'Intersolve-visa';
    const dataUpdate = {
      programFinancialServiceProviderConfigurationName:
        newProgramFinancialServiceProviderConfigurationName,
    };
    const reason = 'automated test';

    // Act
    const response = await updateRegistration(
      programIdPv,
      registrationPvScoped.referenceId,
      dataUpdate,
      reason,
      accessToken,
    );

    // Assert
    expect(response.statusCode).toBe(HttpStatus.BAD_REQUEST);
    expect(response.body).toMatchSnapshot();
  });

  it('should succeed when updating program financial servce provider configuration when missing required properties of new FSP are passed along with the request', async () => {
    // Arrange
    await setupNlrcEnvironment();

    // Intersolve-voucher-whtatsapp does not have e.g. addressStreet which is required for Intersolve-visa
    // The missing attributes can be passed along (or can be updated first)
    const newProgramFinancialServiceProviderConfigurationName =
      'Intersolve-visa';
    const dataUpdate = {
      programFinancialServiceProviderConfigurationName:
        newProgramFinancialServiceProviderConfigurationName,
      addressStreet: 'Teststraat 1',
      addressHouseNumber: '1',
      addressCity: 'Teststad',
      addressPostalCode: '1234AB',
    };
    const reason = 'automated test';

    // Act
    const response = await updateRegistration(
      programIdPv,
      registrationPvScoped.referenceId,
      dataUpdate,
      reason,
      accessToken,
    );

    // Assert
    expect(response.statusCode).toBe(HttpStatus.OK);
    expect(response.body.programFinancialServiceProviderConfigurationName).toBe(
      newProgramFinancialServiceProviderConfigurationName,
    );
  });

  it('should fail when updating program financial servce provider configuration without right permission', async () => {
    // Arrange
    await setupNlrcEnvironment();

    await removePermissionsFromRole(DefaultUserRole.Admin, [
      PermissionEnum.RegistrationFspConfigUPDATE,
    ]);
    accessToken = await getAccessToken();

    // Intersolve-visa and Intersolve-voucher-whatsapp both have 'whatsappPhoneNumber' as required, so this would succeed apart from the permission
    const newProgramFinancialServiceProviderConfigurationName =
      'Intersolve-voucher-whatsapp';
    const dataUpdate = {
      programFinancialServiceProviderConfigurationName:
        newProgramFinancialServiceProviderConfigurationName,
    };
    const reason = 'automated test';

    // Act
    const response = await updateRegistration(
      programIdOcw,
      registrationVisa.referenceId,
      dataUpdate,
      reason,
      accessToken,
    );

    // Assert
    expect(response.statusCode).toBe(HttpStatus.FORBIDDEN);
  });

  it('should succeed updating registration to a newly added FSP config of which the name is not the same as the FSP and doing a payment', async () => {
    // Arrange
    const payment = 1;
    await resetDB(SeedScript.nlrcMultiple);
    await seedIncludedRegistrations(
      [registrationPvScoped],
      programIdPv,
      accessToken,
    );

    const newProgramFinancialServiceProviderConfigurationName =
      'VoucherNumberTwo';
    const fspConfigBody: CreateProgramFinancialServiceProviderConfigurationDto =
      {
        name: newProgramFinancialServiceProviderConfigurationName,
        financialServiceProviderName:
          FinancialServiceProviders.intersolveVoucherWhatsapp,
        label: { en: 'Voucher number 2' },
        properties: [
          {
            name: FinancialServiceProviderConfigurationProperties.password,
            value: 'password',
          },
          {
            name: FinancialServiceProviderConfigurationProperties.username,
            value: 'username',
          },
        ],
      };
    await postProgramFinancialServiceProviderConfiguration({
      programId: programIdPv,
      body: fspConfigBody,
      accessToken,
    });

    const dataUpdate = {
      programFinancialServiceProviderConfigurationName:
        newProgramFinancialServiceProviderConfigurationName,
    };
    const reason = 'automated test';

    // Act
    const response = await updateRegistration(
      programIdPv,
      registrationPvScoped.referenceId,
      dataUpdate,
      reason,
      accessToken,
    );
    await doPayment(
      programIdPv,
      payment,
      15,
      [registrationPvScoped.referenceId],
      accessToken,
    );
    await waitForPaymentTransactionsToComplete(
      programIdPv,
      [registrationPvScoped.referenceId],
      accessToken,
      30_000,
    );

    const transactionsResponse = await getTransactions(
      programIdPv,
      payment,
      registrationPvScoped.referenceId,
      accessToken,
    );

    // Assert
    expect(response.statusCode).toBe(HttpStatus.OK);
    expect(response.body.programFinancialServiceProviderConfigurationName).toBe(
      newProgramFinancialServiceProviderConfigurationName,
    );
    expect(transactionsResponse.text).toContain('success');
    expect(
      transactionsResponse.body[0]
        .programFinancialServiceProviderConfigurationName,
    ).toBe(newProgramFinancialServiceProviderConfigurationName);
    expect(transactionsResponse.body[0].financialServiceProviderName).toBe(
      FinancialServiceProviders.intersolveVoucherWhatsapp,
    );
  });
});
