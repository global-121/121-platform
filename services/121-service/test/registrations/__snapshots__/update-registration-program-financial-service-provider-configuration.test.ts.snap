// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update program financial servce provider configuration of PA should fail when updating program financial servce provider configuration when a required property of new FSP is not yet present 1`] = `
{
  "message": "addressCity: Cannot update 'addressCity' is required for the FSP: 'Intersolve-visa', addressHouseNumber: Cannot update 'addressHouseNumber' is required for the FSP: 'Intersolve-visa', addressPostalCode: Cannot update 'addressPostalCode' is required for the FSP: 'Intersolve-visa', addressStreet: Cannot update 'addressStreet' is required for the FSP: 'Intersolve-visa'",
  "statusCode": 400,
}
`;
